cpp/src/arrow/flight/integration_tests/test_integration.cc

```diff
@@ -410,6 +413,470 @@ class OrderedScenario : public Scenario {
+                   std::optional<Timestamp> expiration_time) {
+    endpoints.push_back(FlightEndpoint{
+        {std::to_string(statuses_.size()) + ": " + ticket}, {}, expiration_time});
+    statuses_.emplace_back(expiration_time);
+  }
+
+  arrow::Result<uint32_t> ExtractIndexFromTicket(const std::string& ticket) {
+    auto index_string = arrow::internal::SplitString(ticket, ':', 2)[0];
+    uint32_t index;
+    if (!arrow::internal::ParseUnsigned(index_string.data(), index_string.length(),
+                                        &index)) {
+      return Status::KeyError("Invalid flight: no index: ", ticket);
+    }
+    if (index >= statuses_.size()) {
+      return Status::KeyError("Invalid flight: out of index: ", ticket);
+    }
+    return index;
+  }
+
+  std::shared_ptr<Schema> BuildSchema() {
+    return arrow::schema({arrow::field("number", arrow::uint32(), false)});
+  }
+
+  std::vector<EndpointStatus> statuses_;
+};
+
+/// \brief The expiration time scenario - DoGet.
+///
+/// This tests that the client can read data that isn't expired yet
+/// multiple times and can't read data after it's expired.
+class ExpirationTimeDoGetScenario : public Scenario {
+  Status MakeServer(std::unique_ptr<FlightServerBase>* server,
+                    FlightServerOptions* options) override {
+    *server = std::make_unique<ExpirationTimeServer>();
+    return Status::OK();
+  }
+
+  Status MakeClient(FlightClientOptions* options) override { return Status::OK(); }
+
+  Status RunClient(std::unique_ptr<FlightClient> client) override {
+    ARROW_ASSIGN_OR_RAISE(
+        auto info, client->GetFlightInfo(FlightDescriptor::Command("expiration_time")));
+    std::vector<std::shared_ptr<arrow::Table>> tables;
+    // First read from all endpoints
+    for (const auto& endpoint : info->endpoints()) {
+      ARROW_ASSIGN_OR_RAISE(auto reader, client->DoGet(endpoint.ticket));
+      ARROW_ASSIGN_OR_RAISE(auto table, reader->ToTable());
+      tables.push_back(table);
+    }
+    // Re-reads only from endpoints that have expiration time
+    for (const auto& endpoint : info->endpoints()) {
+      if (endpoint.expiration_time.has_value()) {
+        ARROW_ASSIGN_OR_RAISE(auto reader, client->DoGet(endpoint.ticket));
+        ARROW_ASSIGN_OR_RAISE(auto table, reader->ToTable());
+        tables.push_back(table);
+      } else {
+        auto reader = client->DoGet(endpoint.ticket);
+        if (reader.ok()) {
+          return Status::Invalid(
+              "Data that doesn't have expiration time "
+              "shouldn't be readable multiple times");
+        }
+      }
+    }
+    // Re-reads after expired
+    for (const auto& endpoint : info->endpoints()) {
+      if (!endpoint.expiration_time.has_value()) {
+        continue;
+      }
+      const auto& expiration_time = endpoint.expiration_time.value();
+      if (expiration_time > Timestamp::clock::now()) {
+        std::this_thread::sleep_for(expiration_time - Timestamp::clock::now());
+      }
```

Comment:
I wonder if this might be flaky in CI (at least in the prior loop), since if it takes more than 2 seconds somehow it'll fail above. Maybe there's no need to test the actual expiration of the ticket? Or, we can treat the expiration as a counter instead of a real expiration time to make things not dependent on the clock.